packages/core/src/amazonqGumby/chat/controller/controller.ts

```diff
@@ -216,10 +222,23 @@ export class GumbyController {
                     this.messenger.sendCompilationInProgress(message.tabID)
                     return
             }
-            CodeTransformTelemetryState.instance.setSessionId()
             this.messenger.sendTransformationIntroduction(message.tabID)
+        } catch (e: any) {
+            // if there was an issue getting the list of valid projects, the error message will be shown here
+            this.messenger.sendErrorMessage(e.message, message.tabID)
+            errorMessage = e.code
+        } finally {
+            // Emit initial metric for /transform
+            const authType = await getAuthType()
+            telemetry.codeTransform_initiateTransform.emit({
```

Comment:
Please see [telemetry.md](https://github.com/aws/aws-toolkit-vscode/blob/master/docs/telemetry.md).
You can remove a lot of this boilerplate code by doing using `telemetry.run()` instead of emit. **Applicable throughout entire PR.**
```
try {
    telemetry.codeTransform_initiateTransform.emit(async () => {
        // logic
        // ...

        telemetry.record({codeTransformSessionId: CodeTransformTelemetryState.instance.getSessionId()})
        
        // some more logic
        // ...
        
        // once you know more information, you can add it. If you knew all information at the start, the one `record` is sufficient
        telemetry.record({credentialSourceId: authType})

        // if something is wrong, instead of recording a string, just throw an exception
        if (authState.amazonQ !== 'connected') {
                void this.messenger.sendAuthNeededExceptionMessage(authState, message.tabID)
                throw new Error('auth-failed') // prefer a more detailed exception, this is just an example.
        }
    })
} catch (e) {
    this.messenger.sendErrorMessage(e.message, message.tabID)
}
```
Result, reason, and reasonDesc will be filled automatically from the appropriate fields of the exception.

Reply:
After taking a look at telemetry.md, I see there are benefit of using run() if the telemetry span across multiple functions or contains a large number of fields that need to be updated. However, looks like for it to set the correct failure result, it would need a `ToolkitError`, and I don't think transform is using ToolkitError everywhere (which may be something to fix, but I'll leave it out of scope for the PR).

I found it easier to read when emitting the whole metric through emit(). It makes it easy to see which fields had been populated, and follows normal coding practices where a metric is emitted at the end of some operation (e.g. try/catch/finally). Since we are not emitting redundant metrics, I'm more inclined to keep using emit() the way it is.